### ISSUE TITLE
`performance-unnecessary-value-param` false positive with iterator as template parameter

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#101181]
```cpp
#include <list>

template<class Iterator>
void func(Iterator it)
{
}

void f()
{
    std::list<int> l;
    func(l.rbegin());
}
```

``` 
<source>:4:20: warning: the parameter 'it' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
    4 | void func(Iterator it)
      |                    ^
      |           const   &
```
https://godbolt.org/z/611TeePzP

**Comment by vbvictor** [comment]
I think that this is not false-positive if we strictly follow check conditions:

> The check is only applied to parameters of types that are expensive to copy which means they are not trivially copyable or have a non-trivial copy constructor or destructor.

`std::reverse_iterator` which is returned from `rbegin` is not trivially copyable and trivially copy constructible
https://godbolt.org/z/5KK7zs171

**Comment by firewave** [comment]
Thanks for pointing that out.

Interestingly it is both when using `libc++` - https://godbolt.org/z/bq5971onb.
